url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/DecodeStageARMv5.java

```diff
@@ -0,0 +1,696 @@
+            default:
+                //未定義
+                //TODO: Not implemented
+                return OpIndex.INS_ARM_UND;
+            }
+        } else if (op == 1) {
+            if (l) {
+                //ldrh
+                return OpIndex.INS_ARM_LDRH;
+            } else {
+                //strh
+                return OpIndex.INS_ARM_STRH;
+            }
+        } else if (op == 2) {
+            if (l) {
+                //ldrsb
+                return OpIndex.INS_ARM_LDRSB;
+            } else {
+                //ldrd
+                return OpIndex.INS_ARM_LDRD;
+            }
+        } else if (op == 3) {
+            if (l) {
+                //ldrsh
+                return OpIndex.INS_ARM_LDRSH;
+            } else {
+                //strd
+                return OpIndex.INS_ARM_STRD;
+            }
+        } else {
+            //未定義
+            //TODO: Not implemented
+            return OpIndex.INS_ARM_UND;
+        }
+    }
+
+    /**
+     * イミディエートのみを取るデータ処理命令、その他の命令をデコードします。
+     *
+     * データ処理イミディエート命令、
+     * ステータスレジスタへのイミディエート移動命令、
+     * の実行
+     *
+     * @param inst ARM 命令
+     * @return 命令の種類
+     */
+    public OpIndex decodeALUImm(InstructionARM inst) {
+        int id = inst.getOpcodeSBitImmID();
+
+        switch (id) {
+        case InstructionARM.OPCODE_S_MSR:
+            return OpIndex.INS_ARM_MSR;
+        case InstructionARM.OPCODE_S_UND:
+            return OpIndex.INS_ARM_UND;
+        default:
+            return decodeALUCommon(inst, id);
+        }
+    }
+
+    /**
+     * その他のデータ処理命令、
+     * をデコードします。
+     *
+     * bit[27:23] = 0b00010
+     * bit[20] = 0
+     *
+     * 各ビットと命令の対応は下記の通りです。
+     *
+     *         | 22  | 21  ||  7  |  6  |  5  |  4  |
+     * --------+-----+-----++-----+-----+-----+-----+
+     * MRS     |  x  |  0  ||  0  |  0  |  0  |  0  |
+     * MSR     |  x  |  1  ||  0  |  0  |  0  |  0  |
+     * BX      |  0  |  1  ||  0  |  0  |  0  |  1  |
+     * CLZ     |  1  |  1  ||  0  |  0  |  0  |  1  |
+     * BLX(2)  |  0  |  1  ||  0  |  0  |  1  |  1  |
+     * BKPT    |  0  |  1  ||  0  |  1  |  1  |  1  |
+     * --------+-----+-----++-----+-----+-----+-----+
+     * QADD    |  0  |  0  ||  0  |  1  |  0  |  1  |
+     * QSUB    |  0  |  1  ||  0  |  1  |  0  |  1  |
+     * QDADD   |  1  |  0  ||  0  |  1  |  0  |  1  |
+     * QDSUB   |  1  |  1  ||  0  |  1  |  0  |  1  |
+     * --------+-----+-----++-----+-----+-----+-----+
+     * SMLAxy  |  0  |  0  ||  1  |  y  |  x  |  0  |
+     * SMLAWxy |  0  |  1  ||  1  |  y  |  0  |  0  |
+     * SMULWxy |  0  |  1  ||  1  |  y  |  1  |  0  |
+     * SMLALxy |  1  |  0  ||  1  |  y  |  x  |  0  |
+     * SMULxy  |  1  |  1  ||  1  |  y  |  x  |  0  |
+     * --------+-----+-----++-----+-----+-----+-----+
+     *
+     * これ以外のパターンは全て未定義命令です。
+     *
+     * @param inst ARM 命令
+     * @return 命令の種類
+     */
+    public OpIndex decodeALUOther(InstructionARM inst) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 30. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.